### PR TITLE
fix(system): credit RAM offload to discrete-GPU nodes on all survey platforms

### LINKED ISSUE #1582: Windows: GPU nodes under-advertise capacity — RAM-offload credit is Linux-gated in the VramBytes survey
**Context.** Sibling of #1566, which #1567 fixed for the `gpus.is_empty()` branch (CPU-only nodes). The GPU-equipped branch has the same Linux gate, and was called out as an out-of-scope follow-up in #1567's description (also re-surfaced in #1571 before it was closed as a duplicate). Opening it as its own issue since it changes the advertised capacity of live GPU nodes mesh-wide and deserves its own validation.

**Mechanism.** In `crates/mesh-llm-system/src/hardware/mod.rs`, non-SoC branch of the `Metric::VramBytes` block:

```rust
#[cfg(target_os = "linux")]
let system_ram = read_system_ram_bytes();
#[cfg(not(target_os = "linux"))]
let system_ram = 0u64;
let ram_offload = system_ram.saturating_sub(vram);
survey.vram_bytes = vram + (ram_offload as f64 * 0.90) as u64;
```

On Linux, a discrete-GPU node advertises `vram + 0.90 × (system_ram − vram)`. On Windows, `system_ram` is hardcoded to `0`, so `ram_offload` is `0` and the node advertises bare VRAM only. (Non-SoC macOS hits the same branch, but Windows is the focus here.)

**Impact, concrete.** My Windows 11 bench (RTX 4070 Ti 12 GB, 32 GB system RAM) advertises 12 GB. The same machine on Linux would advertise 12 + 0.90 × 20 ≈ 30 GB — a ~2.5× under-advertisement. Consequences: the scheduler sizes this node down, split placement treats it as far smaller than it is, and Windows GPU nodes are systematically under-utilized across the mesh.

**Fix direction.** `read_system_ram_bytes()` exists on Windows since #1567, so the likely shape is widening the cfg gate — plus validating that the 0.90 offload factor behaves sensibly with Windows paging before applying it as-is. This is capacity-advertisement behavior, so it warrants measured before/after from a real bench rather than a drive-by cfg change (which is why it stayed out of #1567).

**I can take this** — PR to follow, validated on my bench (advertised capacity before/after + serve behavior).

*Found via code-path analysis on current `main`; measured numbers will come with the PR.*

### PR DESCRIPTION
**Pull request by Virgile-pct**
Fixes #1582

On Windows, a serve node with a discrete GPU budgets bare VRAM locally while the same machine on Linux credits vram + 0.90 x (system_ram - vram). The skippy probe seam hardcoded `system_ram = 0` outside Linux in the non-SoC GPU branch, while the legacy Windows collector path in the same file already applies the offload credit, so the CLI and the serve path disagreed with each other on the same box.

Changes, both in `mesh-llm-system`:

- the non-SoC GPU branch of `apply_gpu_probe_outcome_to_survey` now reads RAM through the seam's injected source instead of a `cfg(target_os = "linux")` block
- the production source is renamed `cpu_only_budget_system_ram` to `survey_system_ram` and carries the platform gating alone (real read on Linux and Windows, zero elsewhere), which keeps the seam platform-pure and the branch testable with injected values
- the 0.90 factor is unchanged, matching both Linux and the legacy Windows path
- five regression tests: nominal credit (12 GB VRAM + 32 GB RAM advertise 30 GB), zero RAM source advertises bare VRAM, RAM below VRAM saturates to bare VRAM, a unified-memory survey never probes RAM, and no probe happens when VramBytes is not requested

Deliberately out of scope: macOS behavior is unchanged by construction (its RAM source stays zero), and the mesh-advertised capacity is intentionally untouched: `mesh_capacity_bytes` keeps advertising per-GPU allocatable VRAM and its comment explicitly keeps the RAM offload budget local-only, so this fix only affects the local admission budget. Also out of scope: three pre-existing Windows clippy findings on untouched code (`download_optional_url`, two `SignalOutcome` variants, one dead field in skippy-runtime), reproduced identically on a clean main checkout, follow-up material.

Validation on the machine from #1566 (RTX 4070 Ti 12 GB, 33.5 GB RAM, Windows 11):

- `cargo test --locked -p mesh-llm-system --lib --features dynamic-native-runtime`: 185 passed on two different Windows machines (this one and a GPU-less laptop)
- `cargo fmt --check`, `repo-consistency no-console-print`, and `cargo clippy --no-deps -p mesh-llm-system --all-targets --features dynamic-native-runtime -- -D warnings` clean apart from the pre-existing findings above
- live before and after, same protocol as #1567 (backport onto v0.76.0-rc8): with a 20 GB model file, stock rc8 fails startup with "local model requires 23.62 GB but this process has 12.88 GB" while `models show` on the same box says "This machine: ~31.4GB available" through the legacy collector; the patched build passes the capacity gate with the expected 31.4 GB budget (12.88 + 0.90 x (33.49 - 12.88)) and proceeds
- the split planner's advertised capacity stays 12.9 GB before and after, confirming the announcement path is untouched

I will address CodeRabbit findings before human review.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved hardware detection for systems with discrete GPUs by accurately accounting for available system RAM that can support GPU workloads.
  * Ensured RAM-offload estimates are applied consistently across CPU-only, discrete-GPU, and unified-memory configurations.
  * Corrected edge cases involving zero or insufficient system RAM and missing GPU memory data.
  * Improved unified-memory detection when related hardware metrics are not requested.

* **Tests**
  * Added coverage for GPU memory and system RAM survey scenarios, including unified-memory devices and omitted metrics.

### CODE REVIEW

File: crates/mesh-llm-system/src/hardware/mod.rs [hunk 1]
```diff
@@ -382,11 +388,7 @@ fn apply_gpu_probe_outcome_to_survey<E>(
             let reserved: u64 = survey.gpu_reserved.iter().flatten().copied().sum();
             survey.vram_bytes = vram.saturating_sub(reserved);
         } else {
-            #[cfg(target_os = "linux")]
-            let system_ram = read_system_ram_bytes();
-            #[cfg(not(target_os = "linux"))]
-            let system_ram = 0u64;
-            let ram_offload = system_ram.saturating_sub(vram);
+            let ram_offload = system_ram().saturating_sub(vram);
```
[the remark this answers is not in the document]
Comment (Virgile-pct) [reply crates/mesh-llm-system/src/hardware/mod.rs:399]: Valid catch, addressed in f955796: the unified-memory decision now comes from the probed gpu facts (`gpus.iter().any(|gpu| gpu.unified_memory)`) instead of `survey.is_soc`, so a VramBytes-only query keeps a unified-memory GPU on the reserved-bytes path. Regression test added for the omission case (VramBytes only, unified GPU, panicking RAM closure).

**Comment by Virgile-pct** [comment]
CodeRabbit's one actionable finding is addressed in f955796 (unified memory decided from the probe rather than from the requested metrics, with a regression test). 186 tests pass on this Windows box; the PR description otherwise stands.

**Review by i386 — CHANGES_REQUESTED** [review]
The production logic and the follow-up unified-memory fix look correct, but the required affected-package suite is not clean on macOS at this head.

`cargo test --locked -p mesh-llm-system --lib --features dynamic-native-runtime` fails reproducibly at `hardware::tests::test_survey_returns_all_metrics` (191 passed, 1 failed): `survey().gpu_name` is `Some("GPU 0")` while the equivalent explicit query returns `None`. I reproduced it twice on `f9557962`, and again after merging current `main` into the temporary review worktree. Exact current `main` passes the same default-parallel suite (186/186), while the PR passes 192/192 only with `--test-threads=1`, so this is a parallel-test interaction exposed by the branch rather than a production-logic objection.

Please make the package suite pass under its normal parallel runner, then rerun the platform workflows. GitHub currently reports all five PR workflows as `action_required`, so there is no CI result that supersedes this local failure yet.

**Comment by Virgile-pct** [comment]
Thanks — reproduced the shape of it, not the failure itself: on Windows without a loaded native runtime the skippy path returns early, so the package suite passes under the default parallel runner here (186/186 on f9557962 and on this head). Analysis instead:

- None of the six tests this PR adds touch shared state: they inject the probe result and the RAM source by parameter.
- `test_survey_returns_all_metrics` exercises the real collector. With the native runtime loaded (your macOS run), that goes through `skippy_devices::gpu_facts()` → `skippy_runtime::backend_devices()` plus the enrichers — process-global state shared with the other real-collector tests (`test_query_*`, `test_detect_collector_*`, the `is_soc` collector tests, and the two Linux skippy-hook tests, which additionally write the probe override). The added tests only shifted scheduling enough for two of those to interleave, which matches what you saw: one enumeration returned a device, the next returned none.
- Fix (ab405cf): the real-collector tests now run in a `serial_test` group (`#[serial(real_collector)]`), same convention as the `#[serial]` env tests in `autoupdate/release_fetch.rs`. The pure seam tests stay parallel. +17 lines, tests only.

Gates on Windows 11: `cargo test --locked -p mesh-llm-system --lib --features dynamic-native-runtime` 186 passed under the default parallel runner; `cargo fmt --check` clean; `repo-consistency no-console-print` clean; `cargo clippy --no-deps -p mesh-llm-system --all-targets --features dynamic-native-runtime -- -D warnings` reports only the two pre-existing Windows-only dead-code errors (`release_fetch.rs:425`, `backend.rs:173`), identical on current `main` and untouched here.

If the interleaving points at `backend_devices()` not being safe to call concurrently, that's a `skippy_runtime` question rather than this PR's — happy to open a follow-up issue with your macOS observation if you'd like it tracked. Could you re-run the parallel suite on your side? The five platform workflows still show `action_required` from here as well.

**Comment by i386** [comment]
Fixed the macOS package-suite failure in b7d2ad94. The failure reproduced even when `test_survey_returns_all_metrics` ran alone, so it was not an inter-test race: `survey()` requests `IsSoc` and `GpuFacts`, while the comparison query omitted both. When the macOS name probe is unavailable, the `GpuFacts` hydration fallback can synthesize `GPU 0` only on the survey side. The test now queries the complete survey contract and compares all returned hardware fields.

Validation at exact head `b7d2ad94` on macOS:
- `cargo test --locked -p mesh-llm-system --lib --features dynamic-native-runtime`: 192 passed, 0 failed under the default parallel runner
- `cargo fmt --all --check`: passed
- `cargo check -p mesh-llm-system --features dynamic-native-runtime`: passed
- `cargo clippy -p mesh-llm-system --all-targets --features dynamic-native-runtime -- -D warnings`: passed
- `cargo check -p mesh-llm`: passed

The broader `cargo clippy -p mesh-llm --all-targets -- -D warnings` still reports 28 pre-existing `unfulfilled_lint_expectations` in `mesh-llm-host-runtime`; none are in this PR or touched package.

**Comment by michaelneale** [comment]
I think this is needed. Never sure if it still makes sense to report one "vram" number - does it? (I don't have alternative proposal but at leaast this helps make it consistent I believe).

**Comment by ndizazzo** [comment]
> Never sure if it still makes sense to report one "vram" number - does it?

@michaelneale IHMO - no... most people avoid offloading to system RAM in the non SoC case, so I think it makes sense to split it out as a discrete unit... however, we don't display it, when it's totally possible to use. Perhaps more resource fields are worth adding for display in the console / CLI, and the user can decide if they want them displayed as "simplified" or broken out into individual line items.

**Comment by Virgile-pct** [comment]
Thanks for b7d2ad94 — you're right, and the isolated run was the test I should have asked for instead of inferring a scheduling interaction from "doesn't reproduce on Windows". The complete-contract comparison is the correct fix; the serial group stays only for the reason your comment gives.

Merged current `main` to clear the conflict with #1591: the `allow(dead_code)` cfg_attr is carried over onto the renamed `survey_system_ram` (still needed: compiled under `test` without `skippy-devices` on macOS and uncalled there). I did not carry the `allow(unused_variables)` onto the renamed closure parameter: this PR uses it in the non-SoC branch on every platform (no cfg on that path), so the lint no longer applies.

Gates on Windows 11 at this head: `cargo test --locked -p mesh-llm-system --lib --features dynamic-native-runtime` 186 passed under the default parallel runner; `cargo fmt --check` clean; `repo-consistency no-console-print` clean; clippy unchanged (only the two pre-existing Windows-only dead-code errors, identical on `main`). The PR/Linux and PR/Quality reds were the llama patch replay (#1587), so a workflow re-run should clear them now that `main` carries the fix.

**Comment by Virgile-pct** [comment]
@michaelneale it stops making sense the moment RAM offload is credited on every platform, which is exactly what this PR does — the single `vram_bytes` now mixes two things with different performance characteristics: device memory and 0.90 × (system RAM − VRAM), which is only a good idea for some workloads. @ndizazzo's split matches what the survey already knows but doesn't surface: per-device VRAM (`gpu_vram`), reserved bytes (`gpu_reserved`), the unified-memory flag, and the offload credit (derivable today only as `vram_bytes − Σ gpu_vram`). Today the two views already disagree for a GPU node: the announced budget is `survey().vram_bytes` (offload credit included — `mesh/node/startup.rs`, gossiped as `vram_gb`), while the console sums per-GPU rated VRAM whenever a GPU list is present (`gpuTotalVramGb` in `status-adapter.ts`) and only falls back to that `vram_gb` otherwise — so the console silently drops the offload credit, and neither number says what it contains. Discovery reports the single `total_vram_bytes`.

Proposal, as a follow-up rather than in this PR: keep `vram_bytes` as the scheduler's single budget for compatibility, expose the breakdown as explicit fields (device VRAM, reserved, RAM-offload credit, unified), and let the console/CLI show either the simplified number or the itemized view, as Nick suggests. That also turns the 0.90 factor from a hidden assumption into a visible, tunable one. Happy to open the issue and take it — I have the Windows bench to measure how the offload credit actually behaves under load, which is where the number is most questionable.

**Comment by Virgile-pct** [comment]
Windows data point at the current head (d3b600d, after the serial_test grouping and your survey-contract fix): `cargo test --locked -p mesh-llm-system --lib --features dynamic-native-runtime` passes 186/186 under the default parallel runner on the RTX 4070 Ti box, `cargo fmt --check` clean. Thanks for the macOS fix and the merge from main.

**Review by i386 — APPROVED** [review]
Verified at head d3b600d8:

**Diff review** — surgical 2-file change in `mesh-llm-system`. The seam refactor is correct: platform gating moves into the production source (`survey_system_ram`), the branch logic stays platform-pure, the 0.90 factor is unchanged, `mesh_capacity_bytes` announcement path untouched. The unified-memory hoist (decided from the probe, not requested metrics) fixes the VramBytes-only regression CodeRabbit flagged. No stale references to the renamed `cpu_only_budget_system_ram`; `serial_test` was already a dev-dependency (no Cargo.toml change needed).

**Local, at head, on macOS (the platform from the earlier review failure)** — `cargo test --locked -p mesh-llm-system --lib --features dynamic-native-runtime`: 192/192 passed under the default parallel runner, including `test_survey_returns_all_metrics`. `cargo +1.97.0 clippy --no-deps -p mesh-llm-system --all-targets --features dynamic-native-runtime -- -D warnings` clean; `cargo +1.97.0 fmt --check` clean.

**CI** — the five PR workflows were stuck at `action_required` (first-time-contributor gate; no CI had run at any head). Approved the runs; all five are green at d3b600d8: macOS, Windows, Linux, Quality, Website.

The changes-requested finding was root-caused (contract-comparison test omitted IsSoc/GpuFacts, not a race) and fixed in b7d2ad94, with live before/after validation on the affected Windows box in the PR thread. The single-`vram_bytes` design question raised in the comments is pre-existing and well-scoped as follow-up.

**Comment by Virgile-pct** [comment]
Opened #1656 for this, with the measurement plan first.
